lib/yolo/models/yolox.ex

```diff
@@ -0,0 +1,184 @@
+defmodule YOLO.Models.Yolox do
+  @moduledoc """
+  YOLOX model implementation for preprocessing input images
+  and postprocessing detections using non-maximum suppression (NMS).
+
+  Supports YOLOX models found at [https://github.com/Megvii-BaseDetection/YOLOX](github.com/Megvii-BaseDetection/YOLOX)
+  """
+
+  @behaviour YOLO.Model
+
+  import Nx.Defn
+
+  @impl true
+  @spec preprocess(YOLO.Model.t(), term(), Keyword.t()) :: {Nx.Tensor.t(), ScalingConfig}
+  def preprocess(model, image, options) do
+    frame_scaler = Keyword.fetch!(options, :frame_scaler)
+    {_, _channels, height, width} = model.shapes.input
+    {image_nx, image_scaling} = YOLO.FrameScalers.fit(image, {height, width}, frame_scaler)
+
+    image_nx =
+      image_nx
+      |> Nx.as_type({:f, 32})
+      # {h, w, c} -> {c, h, w}
+      |> Nx.transpose(axes: [2, 0, 1])
+      # add another axis {3, 640, 640} -> {1, 3, 640, 640}
+      |> Nx.new_axis(0)
+
+    {image_nx, image_scaling}
+  end
+
+  @impl true
+  def precalculate(_model_ref, shapes, options) do
+    {_, _, height, width} = shapes.input
+
+    {grids, expanded_strides} =
+      generate_grids_and_expanded_strides({width, height}, options[:p6] || false)
+
+    %{grids: grids, expanded_strides: expanded_strides}
+  end
+
+  @impl true
+  def postprocess(%{precalculated: precalculated}, model_output, scaling_config, opts) do
+    prob_threshold = Keyword.fetch!(opts, :prob_threshold)
+    iou_threshold = Keyword.fetch!(opts, :iou_threshold)
+    nms_fun = Keyword.get(opts, :nms_fun, &default_nms/3)
+
+    %{grids: grids, expanded_strides: expanded_strides} = precalculated
+
+    model_output
+    |> process_bboxes(grids, expanded_strides)
+    |> nms_fun.(prob_threshold, iou_threshold)
+    |> YOLO.FrameScalers.scale_bboxes_to_original(scaling_config)
+  end
+
+  # YOLOX uses convolutions, so to decode the output, we have to
+  # apply strides to map predictions to input image space.
+  # Translated from https://github.com/Megvii-BaseDetection/YOLOX/blob/d872c71bf63e1906ef7b7bb5a9d7a529c7a59e6a/yolox/utils/demo_utils.py#L139
+  # Generation of grids and expanded strides is split out into `generate_grids_and_expanded_strides`
+  # which is precalculated when the model is loaded, since we don't need to run it for every inference!
+  defn process_bboxes(model_output, grids, expanded_strides) do
```

Comment:
Need to look into whether this could be skipped by exporting onnx with decode_in_inference=True

Reply:
Updated implementation to handle onnx exports with and without head decoding